File: Homework/Ngan_Luu/337-house-robber-iii.py
# 337. https://leetcode.com/problems/house-robber-iii/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not(node):
                return [0,0]
            left=dfs(node.left)
            right=dfs(node.right)
            rob_this_node=node.val+left[1]+right[1]
            not_rob_this_node=max(left)+max(right)
            return [rob_this_node,not_rob_this_node]
        return max(dfs(root))

File: Homework/Duc_Trinh/lesson-7/337-house-robber-III.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

const notRobCache = new Map()
const robCache = new Map()

const getMaxSumOfCurrentNode = (node, isRobbing=true) => {
    if (!node) return 0
    if (isRobbing) {
      if (robCache.has(node)) return robCache.get(node)
      const maxRobVal =  node.val + getMaxSumOfCurrentNode(node.left, false) + getMaxSumOfCurrentNode(node.right, false)
      robCache.set(node, maxRobVal)
      return maxRobVal
    } else {
      if (notRobCache.has(node)) return notRobCache.get(node)
      const maxNotRobVal = Math.max(getMaxSumOfCurrentNode(node.left, false), getMaxSumOfCurrentNode(node.left)) + Math.max(getMaxSumOfCurrentNode(node.right, false), getMaxSumOfCurrentNode(node.right))
      notRobCache.set(node, maxNotRobVal)
      return maxNotRobVal
    }
}

var rob = function(root) {
    return Math.max(getMaxSumOfCurrentNode(root, false), getMaxSumOfCurrentNode(root))
};
